File: backend/schemas/order.py
"""订单模块 Pydantic 数据模型

字段严格对齐前端 order.mock.ts 与 orderApi.ts 契约。
"""
from typing import List, Optional

from pydantic import Field

from .common import SchemaBase
from .enums import OrderStatus


class OrderBase(SchemaBase):
    """订单基础字段"""

    id: str = Field(description="订单ID")
    customer_name: str = Field(description="客户名称")
    date: str = Field(description="订单日期 YYYY-MM-DD")
    total_amount: float = Field(description="订单总金额")
    status: OrderStatus = Field(description="订单状态")
    sender: Optional[str] = Field(default=None, description="发件方")
    receiver: Optional[str] = Field(default=None, description="收件方")
    goods_description: Optional[str] = Field(default=None, description="货物描述")
    shipping_method: Optional[str] = Field(default=None, description="运输方式")
    estimated_delivery: Optional[str] = Field(default=None, description="预计送达日期")


class OrderCreate(SchemaBase):
    """创建订单请求（去掉 id，由后端生成）"""

    customer_name: str = Field(description="客户名称")
    date: str = Field(description="订单日期 YYYY-MM-DD")
    total_amount: float = Field(description="订单总金额")
    status: OrderStatus = Field(description="订单状态")
    sender: Optional[str] = Field(default=None, description="发件方")
    receiver: Optional[str] = Field(default=None, description="收件方")
    goods_description: Optional[str] = Field(default=None, description="货物描述")
    shipping_method: Optional[str] = Field(default=None, description="运输方式")
    estimated_delivery: Optional[str] = Field(default=None, description="预计送达日期")


class OrderUpdate(SchemaBase):
    """更新订单请求（所有字段可选）"""

    customer_name: Optional[str] = None
    date: Optional[str] = None
    total_amount: Optional[float] = None
    status: Optional[OrderStatus] = None
    sender: Optional[str] = None
    receiver: Optional[str] = None
    goods_description: Optional[str] = None
    shipping_method: Optional[str] = None
    estimated_delivery: Optional[str] = None


class OrderResponse(OrderBase):
    """订单响应（继承 OrderBase 全部字段）"""

    pass


class OrderListResponse(SchemaBase):
    """订单列表响应"""

    orders: List[OrderResponse] = Field(description="订单列表")
    total: int = Field(description="总记录数")


class OrderMetricsData(SchemaBase):
    """订单指标统计"""

    total_count: int = Field(description="订单总数")
    in_transit_count: int = Field(description="运输中订单数")
    avg_processing_hours: float = Field(description="平均处理时长(小时)")
    exception_count: int = Field(description="异常订单数")
    total_trend: float = Field(description="总数趋势百分比")
    exception_trend: float = Field(description="异常趋势百分比")


class BatchDispatchRequest(SchemaBase):
    """批量调度请求"""

    order_ids: List[str] = Field(description="待调度订单ID列表")


class BatchDispatchResponse(SchemaBase):
    """批量调度响应"""

    task_id: str = Field(description="调度任务ID")

File: src/services/modules/order.ts
import { httpClient } from '../api/httpClient';

// ================================================================
// 类型定义
// ================================================================

export interface Order {
  id: string;
  customer_name: string;
  date: string;
  total_amount: number;       // 新格式：数值类型（旧格式为 amount: string）
  status: string;
  // 旧格式兼容字段（后端未改造前使用）
  amount?: string;
  status_color?: string;
  // 新增详情字段
  sender?: string;
  receiver?: string;
  goods_description?: string;
  shipping_method?: string;
  estimated_delivery?: string;
}

export interface OrderListResponse {
  orders: Order[];
  total?: number;
}

export interface OrderFilters {
  keyword?: string;
  status?: string;
  dateRange?: string;
}

export interface BatchDispatchResponse {
  task_id: string;
}

/**
 * 订单指标统计数据
 * @description 用于 OrderMetrics 组件顶部 4 个指标卡
 */
export interface OrderMetricsData {
  total_count: number;          // 总订单量
  in_transit_count: number;     // 运输中订单数
  avg_processing_hours: number; // 平均处理时效（小时）
  exception_count: number;      // 异常延误数
  total_trend: number;          // 总订单环比（%）
  exception_trend: number;      // 异常环比（数值变化，负数表示下降）
}

// ================================================================
// 工具函数：兼容旧格式数据
// ================================================================

function normalizeOrder(order: Order): Order {
  // 如果后端返回的是旧格式 amount: string，转换为 total_amount: number
  if (order.amount !== undefined && order.total_amount === undefined) {
    const numStr = String(order.amount).replace(/[$,]/g, '');
    order.total_amount = parseFloat(numStr) || 0;
  }
  return order;
}

// ================================================================
// API
// ================================================================

export const orderApi = {
  // 全局约定：所有接口路径不带尾部斜杠
  // httpClient 请求拦截器会自动去除尾部斜杠，此处保持 /orders 不带斜杠
  getOrders: async (filters?: OrderFilters): Promise<OrderListResponse> => {
    const result = await httpClient.get<OrderListResponse>('/orders', {
      params: filters,
      showLoading: false,
      retry: 2,
    });
    // 兼容旧格式数据
    if (result?.orders) {
      result.orders = result.orders.map(normalizeOrder);
    }
    return result;
  },

  getOrderById: async (orderId: string): Promise<Order> => {
    const order = await httpClient.get<Order>(`/orders/${orderId}`, { showLoading: true });
    return normalizeOrder(order);
  },

  createOrder: (orderData: Partial<Order>) =>
    httpClient.post<Order>('/orders', orderData, { showLoading: true }),

  updateOrder: (orderId: string, orderData: Partial<Order>) =>
    httpClient.put<Order>(`/orders/${orderId}`, orderData, { showLoading: true }),

  deleteOrder: (orderId: string) =>
    httpClient.delete<void>(`/orders/${orderId}`, { showLoading: true }),

  matchCapacity: (orderId: string) =>
    httpClient.post<{ success: boolean; matched: boolean }>(
      `/orders/${orderId}/match-capacity`,
      {},
      { showLoading: true }
    ),

  analyzeCapacity: (orderId: string) =>
    httpClient.get<{ capacity: number; utilization: number }>(
      `/orders/${orderId}/capacity-analysis`,
      { showLoading: true }
    ),

  getCarbonEmission: (orderId: string) =>
    httpClient.get<{ carbon: number; unit: string }>(
      `/orders/${orderId}/carbon`,
      { showLoading: true }
    ),

  batchDispatch: (orderIds: string[]) =>
    httpClient.post<BatchDispatchResponse>('/orders/batch-dispatch', { order_ids: orderIds }, {
      showLoading: true,
    }),

  /**
   * 获取订单指标统计
   * @description 用于 OrderMetrics 组件顶部 4 个指标卡（总订单/运输中/平均时效/异常延误）
   * 后端应基于订单列表实时聚合，保证指标和列表数据一致
   */
  getMetrics: () =>
    httpClient.get<OrderMetricsData>('/orders/metrics', {
      showLoading: false,
    }),
};
